### ISSUE TITLE
Optimal image resizing for classification training

### ISSUE THREAD

**Issue by maxsitt** [ultralytics/yolov5#11701]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi,

the [dataset](https://universe.roboflow.com/maximilian-sittinger/insect_detect_classification/health) I want to use to train a YOLOv5-cls classification model contains images of different sizes and aspect ratios. The objects that should be classified mostly occupy the whole image, so cropping should be avoided to not loose any important visual information. Also the images have a rather low resolution, so downsizing (and maybe also upscaling) should be avoided if possible.

I have several questions regarding the optimal approach for using this image dataset for classification training:
- What `--imgsz` would be optimal in this case? E.g. the maximum width or height occuring in the dataset to avoid downsizing?
- The image resizing/transform seems to happen [here](https://github.com/ultralytics/yolov5/blob/master/utils/augmentations.py#L347-L351). If I understand correctly, the default [`CenterCrop`](https://github.com/ultralytics/yolov5/blob/master/utils/augmentations.py#L373-L383) cuts off the longer sides of the images and then resizes it to the specified `--imgsz`. Using the [`LetterBox`](https://github.com/ultralytics/yolov5/blob/master/utils/augmentations.py#L354-L370) transform instead of CenterCrop should apply padding to all images if the specified `--imgsz` is the maximum width or height occuring in the dataset, so no visual information is lost, is this correct?
- If different amounts of padding are applied dependent on the respective image size, could this somehow lead to a bias during training? E.g. if images of one class are all small with a lot of padding, could the model be trained on this padding instead of the actual object?
- By default the image transformation includes [RGB normalization](https://github.com/ultralytics/yolov5/blob/master/utils/augmentations.py#L351) adapted to the ImageNet dataset. How important is this normalization to get good training results? And how do the ImageNet RGB mean and standard deviation impact this normalization if a different dataset is used?
- Is there a simple way to save the transformed images to a folder, to inspect them and compare different transformation approaches?

Sorry if some answers to these questions might be obvious and thank you for the support!

**Comment by deepConnectionism** [comment]
A 1: The size of `imgsz` should be obtained by experiment, and the general image classification default is 224
A 2: You can use letterbox inference instead:

    return T.Compose([LetterBox(size), ToTensor(), T.Normalize(IMAGENET_MEAN, IMAGENET_STD)])

A 3: I don't think letterbox should be done during training, only during inference 
A 4: normalization can make the model converge faster and obtain a higher acc. In general, if you count the mean and std of your own dataset, rather than using Imagenet, you may be able to bring in a higher acc.
A 5: When training, it will be saved as `train_images.jpg` and you can see the difference.

**Comment by glenn-jocher** [comment]
@deepConnectionism hi,

Thank you for reaching out. Here are some answers to your questions:

1. The optimal `--imgsz` value depends on your specific use case and the dataset you are working with. In general, the default value of `224` is often a good starting point for image classification tasks. However, it is recommended to experiment with different sizes to find the best value for your specific dataset.

2. You are correct. By using the `LetterBox` transform instead of `CenterCrop`, padding will be applied to all images to maintain their aspect ratio without losing any visual information. This is particularly useful when dealing with images of different sizes and aspect ratios, as it ensures that the entire image is preserved.

3. During training, it is generally not recommended to apply letterboxing. This is because the model should learn to classify objects in their original aspect ratios. Letterboxing during training could introduce bias and potentially affect the model's ability to generalize to different aspect ratios during inference.

4. Image normalization is an important step for training deep learning models. The default RGB normalization used in YOLOv5 is adapted to the ImageNet dataset. While it can provide good results in many cases, using the mean and standard deviation specific to your own dataset may improve accuracy. You can calculate the mean and standard deviation of your dataset and replace the ImageNet values with these custom values.

5. To save the transformed images for inspection and comparison, you can modify the code to save the images to a folder during training or inference. You can modify the `save_image` function in the `train.py` or `test.py` script to save the transformed images with different names or in different directories. This way, you can easily compare the effects of different transformations.

I hope this helps! Please let me know if you have any further questions or need any additional assistance.

**Comment by maxsitt** [comment]
Hi @deepConnectionism and @glenn-jocher,

thanks for your answers and tips! I tested and compared different training configurations and some new questions came up:

- In my dataset the median image size is 50x50 px (min: 14 / max: 131). I trained models with `--imgsz 50`, `--imgsz 64` and `--imgsz 128` respectively. Accuracy increased with increasing imgsz, with the highest accuracy achieved when using `--imgsz 128`. When inspecting the `train_images.jpg`, upscaling to 128 px resulted in "better looking" images with more details, compared to upscaling/downscaling to 64 px, even though the resolution of most images is much lower than 128 px. Is upscaling generally recommended compared to downscaling for classification training with different-sized training images? So should the maximum image dimensions be used as reference point to choose `--imgsz` rather than the median image dimensions?

- Using `LetterBox` transform instead of `CenterCrop` resulted in a slightly increased accuracy, probably because in some images important details get cut off if `CenterCrop` is used. However I don't fully understand your answers regarding using `LetterBox` during training. It is not recommended because the model could actually use the padding as feature during training and in this way the resulting model could be biased? E.g. if most images of a specific class get padded during training, but not during inference the accuracy could drop? How to avoid losing visual information during processing of the training images if this is the case?

- Replacing the ImageNet RGB mean and standard deviation with the values of my dataset slightly improved accuracy. That I'm using the `yolov5s-cls.pt` pretrained on ImageNet is actually no problem in this case?

Thanks for some more detailed answers that could help me understand the different approaches better!

**Comment by deepConnectionism** [comment]
@maxsitt 

1. I understand your question is: you have some image data, and these data are different sizes, so you need to choose an `imgsz` to train the model. In general, the big `imgsz` you use when training, the better results you will get. The largest `imgsz` may be related to your dataset. For image classification tasks, you may give a larger input, but the acc may drop. This requires you to experiment to get the suitable imgsz. In addition, there should be no difference between `--imgsz 50` and `--imgsz 64`, because the model requires a multiple of 32, 50 will become 64, please look at the source code, I am not sure, maybe it will become 32.
2. The reason why don't use `LetterBox` during training, instead of using `resize` directly, is that the model can learn more generalized features because the model expects to classify the object correctly even if it has been deformed. Using LetterBox instead of resize in inference is what I understand: this allows the input to maintain the aspect ratio without loss of visual information, meaning that giving the model a simpler input than using `resize` can improve acc. 

> if most images of a specific class get padded during training, but not during inference the accuracy could drop?   

LetterBox is not recommended for training. And using it in inference can improve acc.

> How to avoid losing visual information during processing of the training images if this is the case?   

You can remove the `CenterCrop` operation and go straight to `resize`.

3. No problem, the pre-trained weights just give the model a starting point, and the model will update the weights according to your data.

The above is my personal understanding, there may be mistakes, please refer to carefully.

**Comment by glenn-jocher** [comment]
@deepConnectionism 

1. For your first question, the optimal `--imgsz` value depends on your dataset. In general, using a larger `imgsz` during training can lead to better results. It's recommended to experiment with different sizes to find the most suitable `imgsz` for your specific dataset. Regarding the difference between `--imgsz 50` and `--imgsz 64`, the model requires an input size that is a multiple of 32. Therefore, `--imgsz 50` will be adjusted to `--imgsz 64` internally. You can refer to the source code for more details.

2. During training, it's generally recommended not to use the `LetterBox` transform and instead use the `resize` operation directly. This allows the model to learn more generalized features and perform well even when the objects in the images are deformed. On the other hand, using `LetterBox` during inference helps maintain the aspect ratio of the input without losing any visual information. This can potentially improve accuracy compared to using `resize`. Therefore, the best practice is to avoid using `LetterBox` during training and use it only during inference.

If most images of a specific class are padded during training but not during inference, there could be a drop in accuracy. That's why it's important to ensure consistency in data processing between training and inference. To avoid losing visual information during training, you can remove the `CenterCrop` operation and only use `resize` if necessary.

3. It's perfectly fine to use the `yolov5s-cls.pt` pretrained on ImageNet for your dataset. The pretrained weights provide a starting point for the model, and then the model will adapt its weights according to your specific dataset during training. Replacing the ImageNet RGB mean and standard deviation with the values from your dataset can help improve accuracy since it effectively normalizes the input images based on the statistics of your dataset.

Please note that the above answers are based on my understanding and there may be some nuances. I recommend reviewing the source code and documentation for more detailed information.
